pydantic/main.py

```diff
@@ -283,15 +283,13 @@ def __setattr__(self, name, value):
         elif not self.__config__.allow_mutation:
             raise TypeError(f'"{self.__class__.__name__}" is immutable and does not support item assignment')
         elif self.__config__.validate_assignment:
-            value_, error_ = self.fields[name].validate(value, self.dict(exclude={name}), loc=name)
-            if error_:
-                raise ValidationError([error_], type(self))
-            else:
-                self.__dict__[name] = value_
-                self.__fields_set__.add(name)
-        else:
-            self.__dict__[name] = value
-            self.__fields_set__.add(name)
+            known_field = self.fields.get(name, None)
+            if known_field:
+                value_, error_ = known_field.validate(value, self.dict(exclude={name}), loc=name)
```

Comment:
```suggestion
                value, error_ = known_field.validate(value, self.dict(exclude={name}), loc=name)
```

Reply:
we should have a test that would catch this, please add.

Reply:
Good catch with `value` instead of `value_`!
Mea culpa!

Reply:
Fixed.

Reply:
please can you add a test for this so similar problem doesn't happen again.

Reply:
> we should have a test that would catch *this*, please add.

@samuelcolvin What exactly do you mean by *this*?

We already have `test_validating_assignment_ok()` and `test_validating_assignment_fail()` in `test_validators.py`, and `test_validate_assignment()` and `test_validate_assignment_error()` in `test_dataclasses.py` to check `known_field` behaviour.

I only added tests for unknown extra values.

What scenario do we need to test additionally?

Reply:
Oh, I see.
`value` vs. `value_`.
So we need some validator that changes the value. Like trimming the input, right?

Reply:
yes, I guess something like 

```py
def test_change_assignment():
    class Model(BaseModel):
        a: int

        @validator('a')
        def double_a(cls, v):
            return v * 2

        class Config:
            validate_assignment = True

    m = Model(a=1)
    assert m.a == 2

    m.a = 4
    assert m.a == 8
```

I know it's not technically part of the original change here, but I think if you catch a problem like this by chance it's imperative that you add a test that would automatically catch it next time.

Reply:
Done.